### ISSUE TITLE
Replay loading at Results screen fails after watching an Auto play

### ISSUE THREAD

**Issue by ThePooN** [ppy/osu-stable-issues#314]
osu! version: b20200104.8 (Stable)

Example: https://www.twitch.tv/videos/531954535?t=0h52m8s (happened 4 times on this stream, but there's a timestamp to one of them in this link)

How to reproduce:
- watch a replay
- after it ends, double-click on the next replay
- click watch replay
- if you're unlucky, previously played replay plays instead of the selected replay

Hint for resolution: You can also notice the "now playing" overlay briefly goes to the desired replay for a split-second, before immediately reverting to the previous song.

Edit: Also occured while spectating (not with osu!tourney); previous song loaded and I ended up watching the right player, with the right replay data, on the wrong map.

**Comment by WinterIsPretty** [comment]
Woah so like the cursor was moving right but the circles were in completely different locations??

**Comment by ThePooN** [comment]
Update, 45 days later:  
This bug still occurs. I haven't gone through this bug with spectating (both with osu! and osu!tourney) again, however it still happens while loading replays.

I have 2 new examples (although there were more in this stream) of this on hand, and also with a permalink.  
The "hint for resolution" I mentioned in the OP still applies, however, the game is delayed by 1 second (for casters synchronization) on this VOD, but in reality they are synchronized (this is realtime memory scanning).

Detailed rundown of what I did and the result. Once the bug is triggered (for unknown/random reasons), the behavior I'm describing here is consistent.  
| Timestamp                                                     | Action                                                        | Result                                                                                                                                                                                 |
|---------------------------------------------------------------|---------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| [01:04:33](https://www.twitch.tv/videos/556879495?t=1h04m33s) | Load the replay via double-clicking                           | "Now Playing" updates for a split-second to the desired song, before reverting. The osu! client loads the correct background and song, but displays the previous replay and song data. |
| [01:04:42](https://www.twitch.tv/videos/556879495?t=1h04m42s) | Click "Watch Replay"                                          | Previous replay loads. Background reverts to the previous song. All plays fine, at this point the glitch is over, and we can leave and load the right replays correctly again:         |
| [01:04:56](https://www.twitch.tv/videos/556879495?t=1h04m56s) | Quit the replay with "Esc", and load the correct replay       | Replay loads and plays flawlessly.                                                                                                                                                     |
| ---                                                           | ---                                                           | ---                                                                                                                                                                                    |
| [01:15:35](https://www.twitch.tv/videos/556879495?t=1h15m35s) | Load the replay via double-clicking                           | "Now Playing" updates for a split-second to the desired song, before reverting. The osu! client loads the correct background and song, but displays the previous replay and song data. |
| [01:16:18](https://www.twitch.tv/videos/556879495?t=1h16m18s) | Try loading the same replay via double-clicking again         | Transition works as usual, however the replay fails to load again. BG and song are fine, but all the displayed information are from the previous successfully loaded replay.           |
| [01:16:24](https://www.twitch.tv/videos/556879495?t=1h16m24s) | ^                                                             | ^                                                                                                                                                                                      |
| [01:16:28](https://www.twitch.tv/videos/556879495?t=1h16m28s) | Try loading a previously showcased replay via double-clicking | ^                                                                                                                                                                                      |
| [01:16:30](https://www.twitch.tv/videos/556879495?t=1h16m30s) | Try loading the original replay via double-clicking again     | ^                                                                                                                                                                                      |
| [01:16:38](https://www.twitch.tv/videos/556879495?t=1h16m38s) | Click "Watch Replay"                                          | Previous replay loads. Background reverts to the previous song. All plays fine, at this point the glitch is over, and we can leave and load the right replays correctly again:         |
| [01:17:18](https://www.twitch.tv/videos/556879495?t=1h17m18s) | Quit the replay with "Esc", and load the correct replay       | Replay loads and plays flawlessly.                                                                                                                                                     |

You will notice some unnecessary delays between my actions, that's because I was remotely controlling the streaming machine with extremely bad internet.  
The "Replay has been loaded" messages are triggered when I greedily try loading a replay while one is still playing, this is not related to the bug at all.

**Comment by peppy** [comment]
Just to clarify: The audio and background are from the *new* replay, but the score display and replay is from the *previous* replay?

**Comment by peppy** [comment]
Is there any chance the track is hitting the end just as you load the new track and it's looping? Just trying to brainstorm how this race condition can happen.

**Comment by ThePooN** [comment]
> Just to clarify: The audio and background are from the _new_ replay, but the score display and replay is from the _previous_ replay?

Yes.

> Is there any chance the track is hitting the end just as you load the new track and it's looping? Just trying to brainstorm how this race condition can happen.

Will go beyond just observing and try to find a few ways like that to reproduce it this evening.

**Comment by ThePooN** [comment]
Alright, it's pretty easy to reproduce in the end.  
It doesn't seem to be related to the state of the song, I can consistently reproduce it with the following beatmaps and their respective replays:
- [PSYQUI - Architect (Nifty) [Overqualified]](https://osu.ppy.sh/beatmapsets/874516#taiko/1827736): [Replays](https://github.com/ppy/osu-stable-issues/files/4251364/Architect.zip)
- [Makai Toshi Niigata - Ukiyoe Yokochou (EdamaMe411) [Inner Oni]](https://osu.ppy.sh/beatmapsets/837057#taiko/1752691): [Replays](https://github.com/ppy/osu-stable-issues/files/4251373/Ukiyoe.Yokochou.zip)

I have 3 sequences involving these two beatmaps and 3 replays, one that consistently fails. Their results are consistent.

Sequence One:
Play the Auto replay of Architect. Once at results screen, load the Capu replay of Ukiyoe Yokochou via double-clicking. It will always load the right song and background, but load the previous replay data. When clicking "Watch Replay", the previous replay plays. Trying to load *any* other replay before clicking "Watch Replay" or going back to "Song Select" also consistently fails.

Sequence Two:
Exactly like Sequence One, but instead of loading the replay, enable Auto and play. Same results.

Sequence Three:
Play the ThePooN replay of Architect. Once at results screen, load the Capu replay of Ukiyoe Yokochou via double-clicking. Everything works as intended.

With these sequences, we can guess that loading a replay at Song Select right after watching a replay played with Auto fails.  
This is confirmed on the VOD I linked earlier: I believe every occurrence of this bug happened after watching an Auto replay, and I don't believe I've seen this bug happen in any other case.

TL;DR: Loading a replay at song select after watching a replay with Auto consistently fails.

**Comment by peppy** [comment]
That's precisely what I thought it was going to be, in fact I meant to ask you whether that was the case, but it seemed like if it was, it would be happening *every* time.

I could see how this could happen yesterday and will look at a fix today :).

**Comment by peppy** [comment]
Fixed in cutting edge! Please test the scenarios you've been working with to make sure I haven't broken anything.

**Comment by ThePooN** [comment]
The failing scenarios no longer fail, and it doesn't seem like this introduced any other regression with usual replay loading. :tada:
